File: finding-seams/kotlin/src/main/kotlin/findingseams/schedulevisualising/ScheduleFormatter.kt
package findingseams.schedulevisualising

data class ScheduleSession(
    val start: String? = null,
    val length: Int? = null,
    val title: String? = null,
    val speaker: String? = null,
    val room: String? = null,
    val track: String? = null,
    val type: String? = null,
)

data class ScheduleObject(
    val sessions: List<ScheduleSession>,
)

data class ScheduleFormatOptions(
    val locale: String = "en-US",
    val timezoneOffset: Int = 0,
    val conferenceDay: Int? = null,
    val timezone: String = "UTC",
)

class ScheduleFormatter {
    fun format(scheduleObject: ScheduleObject?, options: ScheduleFormatOptions = ScheduleFormatOptions()): String {
        @Suppress("UNUSED_VARIABLE")
        val _unusedLocale = options.locale

        if (scheduleObject == null) {
            return "No sessions available.\n"
        }

        val sorted = scheduleObject.sessions.sortedWith { a, b ->
            when {
                a.start.isNullOrEmpty() && b.start.isNullOrEmpty() -> 0
                a.start.isNullOrEmpty() -> 1
                b.start.isNullOrEmpty() -> -1
                else -> a.start!!.compareTo(b.start!!)
            }
        }

        val dayLabel = options.conferenceDay?.let { "Day $it" } ?: "Conference"
        val tzLabel = if (options.timezoneOffset == 0) "UTC" else options.timezone

        val sb = StringBuilder()
        sb.append("# Conference Schedule – ").append(dayLabel).append(" (").append(tzLabel).append(")\n\n")

        for (session in sorted) {
            val title = session.title ?: "Untitled Session"
            val speaker = session.speaker?.let { " – $it" } ?: ""
            val room = session.room ?: "TBD"
            val displayStart = shiftTime(session.start ?: "??:??", options.timezoneOffset)
            val displayEnd = session.length?.let { len ->
                shiftTime(addMinutes(session.start ?: "00:00", len), options.timezoneOffset)
            }
            val timeRange = displayEnd?.let { "$displayStart–$it" } ?: displayStart
            val typeTag = session.type?.let { " [$it]" } ?: ""
            sb.append("* ").append(timeRange).append(' ').append(title).append(speaker)
                .append(" @ ").append(room).append(typeTag).append('\n')
        }
        return sb.toString()
    }

    private fun shiftTime(timeStr: String, offsetMinutes: Int): String {
        if (timeStr.isEmpty() || timeStr.contains('?')) return timeStr
        val parts = timeStr.split(':')
        val h = parts.getOrNull(0)?.toIntOrNull() ?: 0
        val m = parts.getOrNull(1)?.toIntOrNull() ?: 0
        val totalMinutes = h * 60 + m + offsetMinutes
        val normalized = ((totalMinutes % (24 * 60)) + 24 * 60) % (24 * 60)
        val adjH = normalized / 60
        val adjM = normalized % 60
        return "%02d:%02d".format(adjH, adjM)
    }

    private fun addMinutes(timeStr: String, minutes: Int): String {
        val parts = timeStr.split(':')
        val h = parts.getOrNull(0)?.toIntOrNull() ?: 0
        val m = parts.getOrNull(1)?.toIntOrNull() ?: 0
        val total = h * 60 + m + minutes
        val endH = (total / 60) % 24
        val endM = total % 60
        return "%02d:%02d".format(endH, endM)
    }
}

File: finding-seams/csharp/FindingSeams/ScheduleVisualising/ScheduleFormatter.cs
namespace FindingSeams.ScheduleVisualising;

public record ScheduleSession(
    string? Start,
    int? Length,
    string? Title,
    string? Speaker,
    string? Room,
    string? Track,
    string? Type);

public record ScheduleObject(IReadOnlyList<ScheduleSession> Sessions);

public record ScheduleFormatOptions(
    string Locale = "en-US",
    int TimezoneOffset = 0,
    int? ConferenceDay = null,
    string Timezone = "UTC");

public class ScheduleFormatter
{
    public string Format(ScheduleObject? scheduleObject, ScheduleFormatOptions? options = null)
    {
        var o = options ?? new ScheduleFormatOptions();
        _ = o.Locale;

        if (scheduleObject is null)
            return "No sessions available.\n";

        var sorted = scheduleObject.Sessions.ToList();
        sorted.Sort((a, b) =>
        {
            if (string.IsNullOrEmpty(a.Start) && string.IsNullOrEmpty(b.Start)) return 0;
            if (string.IsNullOrEmpty(a.Start)) return 1;
            if (string.IsNullOrEmpty(b.Start)) return -1;
            return string.Compare(a.Start, b.Start, StringComparison.Ordinal);
        });

        var dayLabel = o.ConferenceDay is { } d ? $"Day {d}" : "Conference";
        var tzLabel = o.TimezoneOffset == 0 ? "UTC" : o.Timezone;

        var sb = new System.Text.StringBuilder();
        sb.Append("# Conference Schedule \u2013 ").Append(dayLabel).Append(" (").Append(tzLabel).Append(")\n\n");

        foreach (var session in sorted)
        {
            var title = session.Title ?? "Untitled Session";
            var speaker = session.Speaker is { } sp ? $" \u2013 {sp}" : "";
            var room = session.Room ?? "TBD";
            var displayStart = ShiftTime(session.Start ?? "??:??", o.TimezoneOffset);
            var displayEnd = session.Length is { } len
                ? ShiftTime(AddMinutes(session.Start ?? "00:00", len), o.TimezoneOffset)
                : null;
            var timeRange = displayEnd is { } de ? $"{displayStart}\u2013{de}" : displayStart;
            var typeTag = session.Type is { } ty ? $" [{ty}]" : "";
            sb.Append("* ").Append(timeRange).Append(' ').Append(title).Append(speaker)
                .Append(" @ ").Append(room).Append(typeTag).Append('\n');
        }

        return sb.ToString();
    }

    private static string ShiftTime(string timeStr, int offsetMinutes)
    {
        if (string.IsNullOrEmpty(timeStr) || timeStr.Contains('?', StringComparison.Ordinal))
            return timeStr;

        var parts = timeStr.Split(':');
        var h = parts.Length > 0 ? int.Parse(parts[0]) : 0;
        var m = parts.Length > 1 ? int.Parse(parts[1]) : 0;
        var totalMinutes = h * 60 + m + offsetMinutes;
        var normalized = ((totalMinutes % (24 * 60)) + 24 * 60) % (24 * 60);
        var adjH = normalized / 60;
        var adjM = normalized % 60;
        return $"{adjH:D2}:{adjM:D2}";
    }

    private static string AddMinutes(string timeStr, int minutes)
    {
        var parts = timeStr.Split(':');
        var h = parts.Length > 0 ? int.Parse(parts[0]) : 0;
        var m = parts.Length > 1 ? int.Parse(parts[1]) : 0;
        var total = h * 60 + m + minutes;
        var endH = (total / 60) % 24;
        var endM = total % 60;
        return $"{endH:D2}:{endM:D2}";
    }
}
